galsim/wfirst/wfirst_wcs.py

```diff
@@ -0,0 +1,478 @@
+    (+f2) is at a position angle pa_fpa = pa_obsy + 90 + theta_fpa.
+
+    @param PA        Position angle of the observatory Y axis, unless `PA_is_FPA=True`, in which
+                     case it's the position angle of the FPA.  Must be provided as a galsim.Angle.
+    @param world_pos A galsim.CelestialCoord indicating the center of the FPA.
+    @param PA_is_FPA If True, then the position angle that was provided was the PA of the focal
+                     plane array, not the observatory. [default: False]
+    @param SCAs      A single number or iterable giving the SCAs for which the WCS should be
+                     obtained.  If None, then the WCS is calculated for all SCAs.
+                     [default: None]
+    @returns a dict of WCS objects for each SCA.
+    """
+    # Check which SCAs are to be done using a helper routine in this module.
+    SCAs = galsim.wfirst._parse_SCAs(SCAs)
+
+    # Enforce type for PA
+    if not isinstance(PA, galsim.Angle):
+        raise TypeError("Position angle must be a galsim.Angle!")
+
+    # Parse input position
+    if not isinstance(world_pos, galsim.CelestialCoord):
+        raise TypeError("Position on the sky must be given as a galsim.CelestialCoord!")
+
+    # Compute position angle of FPA f2 axis, where positive corresponds to the angle east of North.
+    if PA_is_FPA:
+        pa_fpa = PA
+        pa_obsy = PA - 90.*galsim.degrees - theta_fpa
+    else:
+        pa_obsy = PA
+        pa_fpa = PA + 90.*galsim.degrees + theta_fpa
+    cos_pa = np.cos(pa_fpa.rad())
+    sin_pa = np.sin(pa_fpa.rad())
+
+    # Figure out tangent-plane positions for FPA center:
+    xc_fpa_tp, yc_fpa_tp = _det_to_tangplane_positions(xc_fpa, yc_fpa)
+
+    # Note, this routine reads in the coeffs.  We don't use them until later, but read them in for
+    # all SCAs at once.
+    a_sip, b_sip = _parse_sip_file(sip_filename)
+
+    # Loop over SCAs:
+    wcs_dict = {}
+    for i_sca in SCAs:
+        # Set up the header.
+        header = galsim.FitsHeader()
+        # Populate some necessary variables in the FITS header that are always the same, regardless of
+        # input and SCA number.
+        _populate_required_fields(header)
+
+        # And populate some things that just depend on the overall locations or other input, not on
+        # the SCA.
+        header['RA_TARG'] = (world_pos.ra / galsim.degrees,
+                             "right ascension of the target (deg) (J2000)")
+        header['DEC_TARG'] = (world_pos.dec / galsim.degrees,
+                              "declination of the target (deg) (J2000)")
+        header['PA_OBSY'] = (pa_obsy / galsim.degrees, "position angle of observatory Y axis (deg)")
+        header['PA_FPA'] = (pa_fpa / galsim.degrees, "position angle of FPA Y axis (deg)")
+
+        # Finally do all the SCA-specific stuff.
+        header['SCA_NUM'] = (i_sca, "SCA number (1 - 18)")
```

Comment:
I don't know.  It seems a bit too roundabout to me. Writing the header here, then copying some of them to a dict and then copying them back to a header in the `makeFitsHeader` function.  I'm not sure that's the best way to go.

My thought was that the wcs would just be the regular wcs, and then all the extra header items would be set by `makeFitsHeader`.  But I think I see the logic of your scheme.  Some of the extra keys are intermediate calculations of the wcs construction.  Well, one anyway.  `ORIENTAT` is the `pa_sca` value, which I guess we don't want to have to recalculate.

So given that, I think a simpler approach for both the implementation and the user interface would be the following:
- At the end of this function, after you make `wcs = galsim.GSFitsWCS(header=header)`, just do `wcs.header = header` to save the full header information.
- Then rather than have `galsim.fits.write` look for `image.header`, we could have `wcs.writeToFitsHeader` look for `wcs.header` and copy that over first before doing the class-specific write.  Then the user wouldn't have to do anything at all special to get the right header stuff into the final FITS file.
- We could update the other FitsWCS classes to automatically make this wcs.header field when they read in from a FITS file, so the output files would include any original header info that was in the input file.  (Possibly toggle this behavior with a `save_header` kwarg.)

What do you think?